core/rust/qdbr/src/parquet_write/array.rs

```diff
@@ -0,0 +1,865 @@
+            repetition_levels_byte_length: repetition_levels_byte_length as i32,
+            is_compressed: Some(options.compression != CompressionOptions::Uncompressed),
+            statistics,
+        }),
+    };
+    Ok(DataPage::new(
+        header,
+        buffer,
+        Descriptor {
+            primitive_type,
+            // max level values are ignored and recalculated for group types
+            max_def_level: 0,
+            max_rep_level: 0,
+        },
+        Some(num_rows),
+    ))
+}
+
+// encodes in native QDB format
+pub fn array_to_raw_page(
+    aux: &[[u8; 16]],
+    data: &[u8],
+    column_top: usize,
+    options: WriteOptions,
+    primitive_type: PrimitiveType,
+    encoding: Encoding,
+) -> ParquetResult<Page> {
+    assert!(
+        mem::size_of::<ArrayAuxEntry>() == 16,
+        "size_of(ArrayAuxEntry) is not 16"
+    );
+
+    let num_rows = column_top + aux.len();
+    let mut buffer = vec![];
+    let mut null_count = 0usize;
+
+    let aux: &[ArrayAuxEntry] = unsafe { mem::transmute(aux) };
+
+    let arr_slices: Vec<Option<&[u8]>> = aux
+        .iter()
+        .map(|entry| {
+            let size = entry.size() as usize;
+            let offset = entry.offset() as usize;
+            if size > 0 {
+                assert!(
+                    offset + size <= data.len(),
+                    "Data corruption in ARRAY column"
+                );
+                Some(&data[offset..][..size])
+            } else {
+                null_count += 1;
+                None
+            }
+        })
+        .collect();
+
+    let def_levels_iter =
+        (0..num_rows).map(|i| i >= column_top && arr_slices[i - column_top].is_some());
+    encode_primitive_def_levels(&mut buffer, def_levels_iter, num_rows, options.version)?;
+
+    let definition_levels_byte_length = buffer.len();
+
+    let mut stats = BinaryMaxMinStats::new(&primitive_type);
+
+    match encoding {
+        Encoding::Plain => {
+            encode_raw_plain(&arr_slices, &mut buffer, &mut stats);
+        }
+        Encoding::DeltaLengthByteArray => {
+            encode_raw_delta(&arr_slices, null_count, &mut buffer, &mut stats);
+        }
+        _ => {
+            return Err(fmt_err!(
+                Unsupported,
+                "unsupported encoding {encoding:?} while writing an array column"
+            ));
+        }
+    };
+
+    let null_count = column_top + null_count;
+    build_plain_page(
+        buffer,
+        num_rows,
+        null_count,
+        definition_levels_byte_length,
+        if options.write_statistics {
+            Some(stats.into_parquet_stats(null_count))
+        } else {
+            None
+        },
+        primitive_type,
+        options,
+        encoding,
+    )
+    .map(Page::Data)
+}
+
+fn encode_raw_plain(
+    arr_slices: &[Option<&[u8]>],
+    buffer: &mut Vec<u8>,
+    stats: &mut BinaryMaxMinStats,
+) {
+    for arr in arr_slices.iter().filter_map(|&option| option) {
```

Comment:
this is a duplicate method of that in varchar.encode_plain()

Reply:
I forgot to change bin stats to array stats in case of raw encoding - it makes little sense to collect binary min/max for arrays. So, now these functions are unique.

This is done in 3ca99d267cdc63b2484a73f3b5296a0f483f9dd4